Add reader perspectives and advanced topics

jbeorse opened the pull request:
closes #588 

## What is included in this PR?
Adds four reader perspectives to ODK 2 docs:

- User: Someone using the data management application running on the android tools
- Deployment Architect: An author of a data management application and/or a consumer of collected data.
- System Administrator: Runs the cloud components
- Platform Developer: Makes changes to core android tools

And organizes all of the documentation to fit these perspectives. Readers should have an easier time finding the relevant docs to them.

Also ports the ODK 2 Framework and ODK 2 Sync Protocol into sphinx and integrates them into the docs. ODK 2 Framework is broken into multiple files and spread across perspectives. ODK 2 Sync Protocol is still a massive file that is organized into Platform Developer advanced topics.

Also fixed a bunch of warnings, but introduced more with the Framework and Sync Protocol additions, so the total count is roughly similar.

clarlars commented:
Jeff, this looks great!  This is the first time that I have reviewed the new document structure, and I think it looks so much better.  Thanks for all of your hard work on this!

wbrunette commented on odk2-src/advanced-topics-architect.rst:
```diff
@@ -0,0 +1,15 @@
+Deployment Architect Advanced Topics
+=======================================
+
```
I would repeat your definition of a Deployment Architect here is probably wise since a reader may not have read the intro

wbrunette commented on odk2-src/advanced-topics-developer.rst:
```diff
@@ -0,0 +1,11 @@
+Platform Developer Advanced Topics
+=======================================
+
```
I would repeat your definition of a Platform Developer since a reader may not have read the intro

wbrunette commented on odk2-src/scan-intro.rst:
```diff
@@ -5,7 +5,7 @@ ODK Scan
-  ODK Scan is not yet fully released! It is not guaranteed to work at the same quality level as ODK Survey, Tables, Services, or the rest of the release ODK 2.0 tools.
+  ODK Scan is not yet fully released! It is not guaranteed to work at the same quality level as ODK Survey, Tables, Services, or the rest of the release ODK 2 tools.
 
   It is currently at the **Release Candidate** stage, which mean does not have support for deployments, but the ODK core team does guarantee a migration path (possibly with many tedious manual steps) into future releases.
```
I thought Scan was in Beta not RC

jbeorse replied:
Sure, my mistake. I'll update it.

wbrunette commented on odk2-src/getting-started-2-architect.rst:
```diff
@@ -1,214 +1,39 @@
-- :doc:`suitcase-intro` - a desktop tool for synchronizing data from an ODK 2.0 server so the data can be exported to CSV format.
-
-This page provides a brief end-to-end walk-through of the ODK 2.0 tools. It will cover the following topics:
+This guide is intended for Deployment Architects (see the :ref:`odk-2-perspectives` section for details).
```
I think you should put the Deployment Architect description here and link if they want to find out more. Otherwise people will be constantly cross linking and have a hard time keeping on the page they were on IMHO

jbeorse replied:
Yes I noticed this after the PR. I've got some stuff that will hopefully clarify that I'll add.

linl33 commented:
The new document structure looks very good. 
I have one suggestion, some pages are very long and they can get difficult to read. For example, on the _ODK Tables: Sample Application_ page, I think the document will be easier to read if the 5 demos are each on its own page.

linl33 commented on odk2-src/survey-form-processing.rst:
```diff
@@ -0,0 +1,652 @@
+Two methods:
+
+  - code:`convertSelectionString(linkedModel, selection)`
+  - code:`convertOrderByString(linkedModel, order_by)`
```
there should a semicolon in font of `code`

linl33 commented on odk2-src/config-file-structure.rst:
```diff
@@ -0,0 +1,59 @@
+
+.. _config-structure:
+
+A complimentary description is provided in the user-level documentation. The tool suite stores its configuration and data files on the SDCard in a directory structure rooted at `opendatakit`. The directories underneath this are *application names* and provide isolation of data and configuration from one user-defined application to the next. By default, if not specified, the *application name* is  ...
```
should this be complementary?

linl33 commented on odk2-src/services-managing.rst:
```diff
@@ -0,0 +1,303 @@
+Compatible Servers
+~~~~~~~~~~~~~~~~~~~~~~
+
+The latest version of ODK Services is Version 2.0.3 (February 2018).
```
Maybe we should point users to the releases page instead, so that we won't need to update this every release?

jbeorse commented:
I have addressed each of your comments.

@linl33 

> I have one suggestion, some pages are very long and they can get difficult to read. For example, the ODK Tables: Sample Application page, I think the document will be easier to read if the 5 demos are each on its own page.

I split up the Tables Sample Application and the Survey Application as you recommended. Please take a look and see if it has helped and if you're happy with the new format. There is still the issue that the Tea Houses Demo is quite long, and the Navigating a Form instructions are fairly long, but the situation has improved. 

There is also the Sync Protocol document which is absolutely enormous. But I am not sure where to start on breaking that up. 

Are there any other documents that should be broken up similarly?

linl33 commented:
The new separated samples pages look very good but the sidebar doesn't work for those pages. The samples pages are not on the sidebar. 

There probably isn't an easy way to make the Tea Houses Demo shorter. I think we can just leave it for now. Since those pages are mostly picture, if we put the text beside the pictures instead of above/below them maybe they will be easier to read.

The Sync Protocol page would be difficult be beak up too.

linl33 commented on odk2-src/tables-sample-app.rst:
```diff
@@ -0,0 +1,53 @@
+Prerequisites
+--------------------------------
+
+Install ODK Tables and its prerequisites from the guide:doc:`tables-install`.
```
there should be a space after guide

jbeorse commented:
@linl33 you're right, it looks there can be no subsections in the document if you want the TOC tree to show up in the side bar. I've gone ahead and moved all content in the sample apps into separate documents, and the side bar should work now. 

As for formatting with images, that stuff is all set in templates shared between ODK1 and ODK 2 docs. If we wanted to set up our own that would be a larger discussion about how much we wanted to diverge from the standard style.

I think there could be a larger discussion about how long is "too long" for a document. Hopefully I've got the most egregious ones from this PR though.
